### PR TITLE
PR: Add hanging Indentation to the Editor

### LINKED ISSUE #3480: Editor doesn't support hanging indentation
## Description of your problem

**What steps will reproduce the problem?**
1. In Editor, type 
   `long_function_name(`
   then press Enter.

**What is the expected output? What do you see instead?**
I wish there is an option for hanging indent (of a fixed number of spaces), rather than vertical alignment. That is, I want an option for:

```
# Hanging indent
def long_function_name(
        var_one, var_two, var_three,
        var_four):
    print(var_one)
```

rather than:

```
# Vertical alignment
def long_function_name(
                       var_one, var_two, var_three,
                       var_four):
    print(var_one)
```

Currently the vertical alignment seems to be the only option. But it saves little space when there is a big indentation. PEP also recommends both styles. 

If I could wish for more, I would suggest having the hanging indent of twice the amount for one level (e.g., 4 x 2 = 8 spaces), so as to disambiguate it with subsequent indents that are not continuations, as I copied above from the second example of the PEP: https://www.python.org/dev/peps/pep-0008/#indentation. 

Interestingly, hanging indent (of twice the amount for one level) is also the style of the MATLAB editor, so users converting from MATLAB would find it more familiar. It could be important since Spyder is the IDE of choice for many previous MATLAB users.

**Please provide any additional information below**
## Versions and main components
- Spyder Version: 3.0.0
- Python Version: 2.7.12
- Operating system: Mac OS X Sierra (10.12)
## Dependencies

Please go to the menu entry `Help > Optional Dependencies` (or
`Help > Dependencies`), press the button `Copy to clipboard`
and paste the contents below:

jedi >=0.8.1     :  0.9.0 (OK)
matplotlib >=1.0 :  1.5.3 (OK)
nbconvert >=4.0  :  4.2.0 (OK)
numpy >=1.7      :  1.11.1 (OK)
pandas >=0.13.1  :  0.18.1 (OK)
pep8 >=0.6       :  1.7.0 (OK)
psutil >=0.3     :  4.3.0 (OK)
pyflakes >=0.5.0 :  1.2.3 (OK)
pygments >=2.0   :  2.1.3 (OK)
pylint >=0.25    :  1.6.4 (OK)
qtconsole >=4.2.0:  4.2.1 (OK)
rope >=0.9.4     :  0.9.4 (OK)
sphinx >=0.6.6   :  1.4.1 (OK)
sympy >=0.7.3    :  1.0 (OK)

### PR DESCRIPTION
**Pull request by yulkang**
Fixes #3480 

----

This commit has an option for hanging indent at menu - python - Preferences - Editor - Advanced settings - Hanging indentation. It is unchecked by default, but when checked, allows hanging indentation, as follows:

```
def long_function_name(
        var1, var2):
    print("This works. Yeah!")

if long_function_name(
        var1, var2):
    print("also works")
```

The case I found not working is the following, where the subsequent line is indented one step too much:

```
if long_array_name[
        var1]:
            print("this does not work yet")
```

But such cases are rarer for me, so it would be useful even as it stands now. I haven't found why this happens, and I would appreciate help.

I checked the following and they work as expected, too:
- the new option is saved after quitting, and
- the default behavior, vertical alignment, is recovered when the new option is unchecked.

Please let me know if there are other issues.

### CODE REVIEW

**Comment by yulkang** [comment]
@ccordoba12 - Will do. I think I will be able to work on this tonight EST. I will remove the options and probably look at the minor bug after "[" as well.

**Comment by jitseniesen** [comment]
I'm happy to act as reviewer.

@yulkang Can I ask whether if the user types any text after the opening parenthesis, they will still get a vertical indent and not a hanging indent? As in:

```
def long_function_name(var_one, var_two, var_three,
                       var_four):
    print(var_one)
```

**Comment by yulkang** [comment]
@jitseniesen No, the user will get a hanging indent, as the first example in my first post.  I'd prefer always having hanging indents for the sake of space.  I can imagine other people preferring vertical alignments, but it seems that adding another option is burdensome, too.

I removed the options from the branch and it works as described in the first post, so I think you can test it now.

**Comment by jitseniesen** [comment]
So this patch will change the behaviour into:
```
def long_function_name(var_one, var_two, var_three,
        var_four):
    print(var_one)
```

This is against PEP8 standards as I understand them. I don't like it that Spyder would make it easier to write non-PEP8-compliant code than compliant code.  @ccordoba12 Is this behaviour what you want?

**Comment by yulkang** [comment]
@jitseniesen I think it is consistent with PEP8:

https://www.python.org/dev/peps/pep-0008/#indentation

The behavior is the same as the second example in the linked section.  I copied it below:

# Indentation
Use 4 spaces per indentation level.

Continuation lines should align wrapped elements either vertically using Python's implicit line joining inside parentheses, brackets and braces, or using a hanging indent [7] . When using a hanging indent the following should be considered; there should be no arguments on the first line and further indentation should be used to clearly distinguish itself as a continuation line.

Yes:

```
# Aligned with opening delimiter.
foo = long_function_name(var_one, var_two,
                         var_three, var_four)

# More indentation included to distinguish this from the rest.
def long_function_name(
        var_one, var_two, var_three,
        var_four):
    print(var_one)

# Hanging indents should add a level.
foo = long_function_name(
    var_one, var_two,
    var_three, var_four)

```

**Comment by jitseniesen** [comment]
The bit I am worried about is the phrase "there should be no arguments on the first line" in the text you copied.

**Comment by yulkang** [comment]
I guess that could be left to the programmer. I don't think an IDE should enforce all aspects of the style. I would like it as long as it helps.

**Comment by yulkang** [comment]
Or perhaps the behavior could be conditionalized, so that hanging indent is used only when enter is pressed right after the opening parenthesis, and vertical alignment is used otherwise.  

That would be more complicated, but I think it makes sense since there is no saving of space with vertical alignment when enter is pressed right after the opening parenthesis, and since using hanging indent in other cases is against PEP8.  @jitseniesen, @ccordoba12 - would that be fine?

**Comment by jitseniesen** [comment]
> Or perhaps the behavior could be conditionalized, so that hanging indent is used only when enter is pressed right after the opening parenthesis, and vertical alignment is used otherwise. 

As far as I am concerned, that would be perfect.

**Comment by goanpeca** [comment]
@yulkang could your rebase or merge to resolve conflicts?

Also any updates based on the 

>Or perhaps the behavior could be conditionalized, so that hanging indent is used only when enter is pressed right after the opening parenthesis, and vertical alignment is used otherwise.
As far as I am concerned, that would be perfect.

I agree with @jitseniesen the hanging indent makes sense only if no args are present after the parenthesis.

Thanks :-)

**Comment by yulkang** [comment]
@jitseniesen, sorry for the delay! Now it works as following:
```
def function_name(var1,
                  var2):
    print('works')
    
def function_name2(
        hanging_var1,
        hanging_var2):
    print('works, too')
    
if function_name(var1,
                 var2):
    print('works')
    
if function_name2(
        hanging_var1,
        hanving_var2):
    print('works')

if array_name[
        hanging_ix1,
        hanging_ix2]:
    print('works')
    
if array_name[ix1,
              ix2]:
    print('works')
    
if dict_name{
        hanging_key1:val1,
        hanging_key2:val2}:
    print('breaks')
    
if dict_name{key1:val1,
             key2:val2}:
    print('works')
```

@goanpeca  Thank you for letting me know! I will check what the conflicts are.

**Comment by goanpeca** [comment]
hi you added some project files that shoild not have been included

**Comment by yulkang** [comment]
@goanpeca Sorry! Is it fine now?

**Comment by goanpeca** [comment]
@yulkang looking good :-)

**Comment by jitseniesen** [comment]
@yulkang Look at the file `spyder/widgets/sourcecode/tests/test_autoindent.py` [(link)](https://github.com/spyder-ide/spyder/blob/master/spyder/widgets/sourcecode/tests/test_autoindent.py) for inspiration for how to write tests. The page https://github.com/spyder-ide/spyder/wiki/Dev:-Testing describes the testing system that we use. I would like to see a test for the hanging indent; it shouldn't be too hard.

Actually, the last test is pretty much what you want. It is currently marked `xfail`, which means that the test is failing with the current version of Spyder, so if you can change it in a working test that would be progress.

While playing around, I discovered that the hanging indent does not work if you are at the very first line of the file; can you have a look at that too please?

**Comment by rlaverde** [comment]
> While playing around, I discovered that the hanging indent does not work if you are at the very first line of the file; can you have a look at that too please?

indentation doesn't work at the first line (not just hanging indent) #3680, and I already fix  that in #3663

**Comment by yulkang** [comment]
@jitseniesen Thank you for the review! I have something urgent to work on until Monday, and I will be able to work on the changes afterwards. @rlaverde Thank you for pointing that out!

**Comment by yulkang** [comment]
@jitseniesen I implemented the suggested changes and fixed/added checks regarding hanging indent, all of which works for me now. (There were several checks that were outdated because they didn't use hanging indent.) However, there seem to be conflicts now, and I can't see how to solve them. Can you help me with this?

I tried several strategies, all of which failed: 
(A) I tried merging the new commits with the 3.x branch myself, but I couldn't test it because my environment lacked some modules (I forgot which... I can check if necessary). 
(B) I also tried merging with the commit 07bedec3 (where 3.x merged with #3663), and then pushing, but it also produced conflicts.
(C) I finally tried to craft the changes in a way that avoids conflicts. That is the current version pushed. But apparently the attempt was unsuccessful...

**Comment by jitseniesen** [comment]
@yulkang I guess the files you worked on have changed in the mean time and that is what's causing your troubles. An alternative is to rebase, as explained here: https://github.com/spyder-ide/spyder/wiki/Dev:-Github-Workflow

**Comment by jitseniesen** [comment]
@yulkang Please let me know when you have finished. It is quite difficult for me to see where you are now because of all the merges.

You need to change the target branch for this pull request from `master` (which will become version 4.0) to `3.x` so that your changes can go into version 3.1. See https://help.github.com/articles/changing-the-base-branch-of-a-pull-request/ for instructions.

If you are stuck with the Git merging business, I'll be happy to do it for you, as it is partially my fault with the long time I took to review your work. I think if you give me permission to edit the pull request following the instructions at https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/ (but I have never done this). Alternatively, I can open a new pull request based on your changes.

**Comment by yulkang** [comment]
> If you are stuck with the Git merging business, I'll be happy to do it for you

@jitseniesen Thank you, that would be very helpful! I changed the base branch to 3.x and simplified changes so that there is no conflict, but somehow some checks are failing. I allowed edits from maintainers, and I would appreciate it a lot  if you can help me with the checks.

File: spyder/widgets/sourcecode/codeeditor.py [hunk 1]
```diff
@@ -1681,24 +1681,37 @@ def fix_indent(self, forward=True, comment_or_string=False):
         block_nb = cursor.blockNumber()
         # find the line that contains our scope
         diff = 0
```
Comment (yulkang) [inline spyder/widgets/sourcecode/codeeditor.py]: Done!

File: spyder/widgets/sourcecode/codeeditor.py [hunk 2]
```diff
@@ -1681,24 +1681,37 @@ def fix_indent(self, forward=True, comment_or_string=False):
...
         prevline = None
         for prevline in range(block_nb-1, -1, -1):
             cursor.movePosition(QTextCursor.PreviousBlock)
             prevtext = to_text_string(cursor.block().text()).rstrip()
             if (self.is_python_like() and not prevtext.strip().startswith('#') \
               and prevtext) or prevtext:
-                if prevtext.strip().endswith(')'):
+                if prevtext.strip().endswith(')') \
+                        or prevtext.strip().endswith(']') \
+                        or prevtext.strip().endswith('}'):
```
Comment (yulkang) [inline spyder/widgets/sourcecode/codeeditor.py]: This is done, too.

**Comment by ccordoba12** [comment]
@yulkang, thanks a lot for your work on this. Unfortunately, it seems your changes are breaking previous work of indentation using tabs.

@rlaverde, could you review @yulkang's work and let him what changes are interfering with your work? Thanks :-)

**Comment by yulkang** [comment]
@jitseniesen Thank you for the fix! Now I see that fix_indent() caused the problems. I thought it simulates the situation where the given text is typed, but perhaps I was wrong if it is supposed to work on the last line only. Just out of curiosity, what situation is it intended to simulate, then?

**Comment by jitseniesen** [comment]
I think the idea of fix_indent() is to check only the indentation of the last line.

However, I think there is a bit of a problem with nested parentheses. The third line is indented incorrectly:

```python
if function_name(
        another_function(var),
                        var2):
    pass
```

**Comment by ccordoba12** [comment]
> However, I think there is a bit of a problem with nested parentheses

Balanced parentheses should be removed from the computation, but that's something I assigned @rlaverde to work on.

@jitseniesen, is this ready to me merged?

**Comment by jitseniesen** [comment]
@ccordoba12 Yes, it's ready to be merged

**Comment by ccordoba12** [comment]
Thanks for your work @yulkang, and @jitseniesen for reviewing it :-)

It's a great improvement over what we had!

**Comment by rlaverde** [comment]
Sorry for the late reply, I review it and It's a nice improvement :-)

I'm working in balanced parentheses, I'll write some falling test and open a PR

**Comment by yulkang** [comment]
Thank you @ccordoba12, @jitseniesen, and @rlaverde so much, for being patient and helpful—I learned a lot in the process!

**Comment by ccordoba12** [comment]
@yulkang, great to hear that! You're more than welcome to keep sending PRs to improve Spyder :-)

**Pull request merged**